File: java/ProductSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

public class ProductSum {

    public static List<List<Integer>> productSum(List<Long> arr, long target) {
        final List<List<Integer>> comb = new ArrayList<>();
        final List<Integer> skiped_idxs = new ArrayList<>();
        IntStream.range(0, arr.size())
                .forEach(i -> {
                    final long curr_val = arr.get(i);
                    skiped_idxs.add(i);
                    backctrack(arr, target, curr_val, comb, skiped_idxs);
                    skiped_idxs.clear();
                });
        return comb;
    }

    private static void backctrack(
            List<Long> arr,
            long target,
            long curr_val,
            List<List<Integer>> comb,
            List<Integer> skiped_idxs) {

        final List<Integer> pos = new ArrayList<>(skiped_idxs);
        IntStream.range(0, arr.size())
                .forEach(i -> {
                    if (skiped_idxs.contains(i))
                        return;

                    long val = arr.get(i);
                    long sum = val + curr_val;

                    if (sum == target) {
                        pos.add(i);
                        comb.add(new ArrayList<>(pos));
                        pos.remove(pos.size() - 1);
                    } else if (sum < target) {
                        skiped_idxs.add(i);
                        backctrack(arr, target, sum, comb, skiped_idxs);
                        skiped_idxs.remove(pos.size() - 1);
                    }
                });
    }

    public static void main(String[] args) {
        final List<Long> vals = Arrays.asList(1L, 4L, 2L, 4L, 5L, 1L);
        final List<Long> arr = new ArrayList<>(vals);
        final long pres = 6;

        productSum(arr, pres).forEach(a -> {
            System.out.print("[ ");
            for (long val : a) {
                System.out.printf("%d ", val);
            }
            System.out.println("]");
        });
    }
}

File: src/day1/product_sum.rs
pub fn product_sum(arr: &[usize], target: usize) {
    let mut combs: Vec<Vec<usize>> = Vec::new();
    let mut hasht: Vec<usize> = Vec::new();

    arr.iter().enumerate().for_each(|(i, &v)| {
        hasht.push(i);
        backtrack(arr, target, v, &mut combs, &mut hasht);
        hasht.pop();
    });
    println!("{combs:?}")
}

fn backtrack(
    arr: &[usize],
    target: usize,
    curr_val: usize,
    combs: &mut Vec<Vec<usize>>,
    hasht: &mut Vec<usize>,
) {
    for (i, &v) in arr.iter().enumerate() {
        if hasht.contains(&i) {
            continue;
        }

        let sum = curr_val + v;
        if sum == target {
            hasht.push(i);
            combs.push(hasht.clone());
            hasht.pop();
        } else if sum < target {
            hasht.push(i);
            backtrack(arr, target, sum, combs, hasht);
            hasht.pop();
        }
    }
}
